forge/src/executor/inspector/cheatcodes/mod.rs

```diff
@@ -0,0 +1,300 @@
+/// Cheatcodes related to the execution environment.
+mod env;
+pub use env::{Prank, RecordAccess};
+/// Assertion helpers (such as `expectEmit`)
+mod expect;
+pub use expect::{ExpectedEmit, ExpectedRevert};
+/// Cheatcodes that interact with the external environment (FFI etc.)
+mod ext;
+/// Cheatcodes that configure the fuzzer
+mod fuzz;
+/// Utility cheatcodes (`sign` etc.)
+mod util;
+
+use self::expect::{handle_expect_emit, handle_expect_revert};
+use crate::{abi::HEVMCalls, executor::CHEATCODE_ADDRESS};
+use bytes::Bytes;
+use ethers::{
+    abi::{AbiDecode, AbiEncode},
+    types::Address,
+};
+use revm::{
+    opcode, CallInputs, CreateInputs, Database, EVMData, Gas, Inspector, Interpreter, Return,
+};
+use std::collections::BTreeMap;
+
+/// An inspector that handles calls to various cheatcodes, each with their own behavior.
+///
+/// Cheatcodes can be called by contracts during execution to modify the VM environment, such as
+/// mocking addresses, signatures and altering call reverts.
+#[derive(Default)]
+pub struct Cheatcodes {
+    /// Whether FFI is enabled or not
+    ffi: bool,
+
+    /// Address labels
+    pub labels: BTreeMap<Address, String>,
+
+    /// Prank information
+    pub prank: Option<Prank>,
+
+    /// Expected revert information
+    pub expected_revert: Option<ExpectedRevert>,
+
+    /// Recorded storage reads and writes
+    pub accesses: Option<RecordAccess>,
+
+    /// Mocked calls
+    pub mocked_calls: BTreeMap<Address, BTreeMap<Bytes, Bytes>>,
+
+    /// Expected calls
+    pub expected_calls: BTreeMap<Address, Vec<Bytes>>,
+
+    /// Expected emits
+    pub expected_emits: Vec<ExpectedEmit>,
+}
+
+impl Cheatcodes {
+    pub fn new(ffi: bool) -> Self {
+        Self { ffi, ..Default::default() }
+    }
+
+    fn apply_cheatcode<DB: Database>(
+        &mut self,
+        data: &mut EVMData<'_, DB>,
+        caller: Address,
+        call: &CallInputs,
+    ) -> Result<Bytes, Bytes> {
+        // Decode the cheatcode call
+        let decoded = HEVMCalls::decode(&call.input).map_err(|err| err.to_string().encode())?;
+
+        // TODO: Log the opcode for the debugger
+        env::apply(self, data, caller, &decoded)
+            .or_else(|| util::apply(self, data, &decoded))
+            .or_else(|| expect::apply(self, data, &decoded))
+            .or_else(|| fuzz::apply(data, &decoded))
+            .or_else(|| ext::apply(self.ffi, &decoded))
+            .ok_or_else(|| "Cheatcode was unhandled. This is a bug.".to_string().encode())?
```

Comment:
do we know the efficiency loss here by repeatedly trying & failing until we match? i.e. 5 match statements vs 1 match statement? I don't imagine its that bad but just curious if thats a concern here